OpenRA.Mods.Common/Traits/Detonates.cs

```diff
@@ -0,0 +1,130 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using System.Text;
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Orders;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class DetonatesInfo : ITraitInfo, Requires<RenderUnitInfo>
+	{
+		[Desc("Measured in ticks.")]
+		public readonly int ChargeDelay = 0;
+		public readonly string ChargeSequence = null;
+		public readonly string ChargeSound = null;
+
+		[Desc("Measured in ticks.")]
+		public readonly int DetonationDelay = 0;
+		[WeaponReference]
+		public readonly string DetonationWeapon = null;
+
+		public readonly string OrderVoice = "Attack";
+
+		public readonly bool KillsSelf = true;
+
+		public object Create(ActorInitializer init) { return new Detonates(init.Self, this); }
+	}
+
+	public class Detonates : IIssueOrder, IResolveOrder, IOrderVoice
+	{
+		readonly Actor self;
+		readonly DetonatesInfo info;
+		readonly RenderUnit renderUnit;
+		public bool Deployed;
+
+		public Detonates(Actor self, DetonatesInfo info)
+		{
+			this.self = self;
+			this.info = info;
+			renderUnit = self.Trait<RenderUnit>();
+		}
+
+		public IEnumerable<IOrderTargeter> Orders
+		{
+			get
+			{
+				yield return new TargetTypeOrderTargeter(new[] { "DetonateAttack" }, "DetonateAttack", 5, "attack", true, false) { ForceAttack = false };
+				yield return new DeployOrderTargeter("Detonate", 5);
+			}
+		}
+
+		void Detonate()
+		{
+			self.World.AddFrameEndTask(w =>
+			{
+				if (info.DetonationWeapon != null)
+				{
+					var weapon = self.World.Map.Rules.Weapons[info.DetonationWeapon.ToLowerInvariant()];
+
+					// Use .FromPos since this actor is killed. Cannot use Target.FromActor
+					weapon.Impact(Target.FromPos(self.CenterPosition), self, Enumerable.Empty<int>());
+				}
+
+				if (info.KillsSelf)
+					self.Kill(self);
+			});
+		}
+
+		protected void StartDetonationSequence()
+		{
+			if (Deployed)
+				return;
+
+			Deployed = true;
+
+			var chargeOverlay = self.TraitsImplementing<INotifyCharging>();
+			foreach (var notify in chargeOverlay)
+				notify.Charging(self, Target.FromActor(self));
+
+			if (info.ChargeSequence != null)
+				renderUnit.PlayCustomAnimationRepeating(self, info.ChargeSequence);
+
+			self.QueueActivity(new Wait(info.ChargeDelay, false));
+			self.QueueActivity(new CallFunc(() => Sound.Play(info.ChargeSound, self.CenterPosition)));
+			self.QueueActivity(new Wait(info.DetonationDelay, false));
+			self.QueueActivity(new CallFunc(Detonate));
+		}
+
+		public void ResolveOrder(Actor self, Order order)
+		{
+			if (order.OrderString == "DetonateAttack")
```

Comment:
You should probably check Deployed (rename to IsDeployed) here and return if true.